File: php/113-path-sum2.php
<?php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     public function __construct($value) 
    { $this->val = $value; }
 * }
 */
class Solution 
{
    private $paths = [];
    /**
     * @param TreeNode $root
     * @param Integer $sum
     * @return Integer[][]
     */
    public function pathSum($root, $sum) 
    {
        if ($root == null) {
            return $this->paths;
        }
        $this->DFS($root, $sum);
        return $this->paths;
    }
    public function DFS($node, $sum, $path = []) 
    {
        if ($node == null) {
            return ;
        }
        $path[] = $node->val;
        if ($node->left == null && $node->right == null) {
            if (array_sum($path) == $sum) {
                $this->paths[] = $path;
            }
            return ;
        }
        $this->DFS($node->left, $sum, $path);
        $this->DFS($node->right, $sum, $path);
    }
}

File: go/113-path-sum2.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var paths [][]int = [][]int{}

func pathSum(root *TreeNode, sum int) [][]int {
	if root == nil {
		return [][]int{}
	}
	paths = paths[0:0]
	getAllPath(root, sum, []int{})
	return paths
}
func getAllPath(root *TreeNode, sum int, path []int) {
	if root == nil {
		return
	}
	if root.Val == sum && root.Left == nil && root.Right == nil {
		paths = append(paths, append(path, root.Val))
		return
	}
	target := sum - root.Val
	p := make([]int, len(path))
	copy(p, path)
	p = append(p, root.Val)
	if root.Left != nil {
		getAllPath(root.Left, target, p)
	}
	if root.Right != nil {
		getAllPath(root.Right, target, p)
	}
}
